netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -234,11 +262,27 @@ private void onHeadersRead(int streamId, Http2Headers headers, boolean endStream
   /**
    * Handler for an inbound HTTP/2 DATA frame.
    */
-  private void onDataRead(int streamId, ByteBuf data, boolean endOfStream) {
+
+  private void onDataRead(int streamId, ByteBuf data, int padding, boolean endOfStream) {
+    if (!pinging) {
+      pinging = true;
+      sendDataPing(ctx());
+    }
+    int currentDataSize = getDataSizeSincePing();
+    setDataSizeSincePing(currentDataSize + data.readableBytes() + padding);
     NettyClientStream stream = clientStream(requireHttp2Stream(streamId));
     stream.transportDataReceived(data, endOfStream);
   }
 
+  private void sendDataPing(ChannelHandlerContext ctx) {
```

Comment:
@mfcripps this method and the others could also be encapsulated in that inner class.